### PR TITLE
feat: SpiceDB chart operator parity features and production-ready enhancements

### PR DESCRIPTION
**Pull request by salekseev**
# SpiceDB Helm Chart: Operator Parity & Production Enhancements

This PR brings the SpiceDB Helm chart to **true operator parity** by matching all spicedb-operator defaults and capabilities, while adding production-ready enhancements and comprehensive documentation improvements.

## 🎯 Primary Goals

1. **True Operator Parity** - Match ALL spicedb-operator defaults and capabilities (validated against source)
2. **Production Readiness by Default** - HA configuration out of the box (2 replicas, dispatch enabled)
3. **Documentation Excellence** - Comprehensive, well-organized documentation
4. **Code Quality** - Clean, maintainable codebase with explicit validation

## 🚨 Breaking Changes (v2.0.0)

**Validated against operator source code** ([`pkg/config/config.go`](https://github.com/authzed/spicedb-operator/blob/main/pkg/config/config.go)):

### 1. replicaCount: 1 → 2 (Operator Parity)
- **Old**: Single replica (not HA)
- **New**: 2 replicas by default (basic HA, matches operator)
- **Impact**: Existing deployments scale to 2 replicas on upgrade
- **Rationale**: Operator defaults to HA configuration
- **Migration**: Set `replicaCount: 1` to maintain single replica, or use `values-presets/development.yaml`

### 2. dispatch.enabled: false → true (Operator Parity)
- **Old**: Dispatch disabled by default
- **New**: Dispatch enabled by default with kubernetes:// service discovery (matches operator)
- **Impact**: Inter-pod dispatch communication enabled on upgrade
- **Rationale**: Operator enables dispatch for all persistent datastores with kubernetes:// scheme
- **Technical Details**:
  - Uses `kubernetes:///service.namespace:dispatch` format (gRPC kuberesolver)
  - Port name resolution (`dispatch`) instead of port number (`50053`)
  - Requires RBAC permissions: `get`, `list`, `watch` on endpoints (included in chart)
  - Consistent hash load balancing across pods
- **Migration**: Set `dispatch.enabled: false` to disable, or use `values-presets/development.yaml`

### 3. migrations.logLevel: info → debug (Operator Parity)
- **Old**: Info-level migration logging
- **New**: Debug-level migration logging (matches operator)
- **Impact**: More verbose migration logs
- **Rationale**: Operator uses debug for detailed migration troubleshooting
- **Migration**: Set `migrations.logLevel: info` for less verbose logs

### 4. config.autogenerateSecret: true → false (Security Best Practice)
- **Old**: Auto-generate secrets by default
- **New**: Require explicit opt-in for auto-generation
- **Impact**: Requires `config.existingSecret` or explicit `config.autogenerateSecret: true`
- **Rationale**: Security best practice - prevent accidental use of auto-generated secrets in production
- **Migration**: Use `values-presets/development.yaml` (enables auto-generation) or provide `existingSecret`

### 🔄 Migration System Enhancement
- **Automatic migration tracking** via ConfigMap with migration hash calculation
- **Pre-upgrade validation hooks** to prevent unsafe migrations (e.g., datastore engine changes)
- **Explicit secret validation** - Fail fast when non-memory datastore configured without secret
- **Migration job cleanup** with post-upgrade hooks
- **Operator-compatible annotations** on all resources for migration support
- **113 comprehensive migration hook tests** (100% passing)

### 🔀 Dispatch Cluster Configuration (kubernetes:// Service Discovery)
- **Kubernetes-native service discovery** using gRPC kuberesolver
- **Port name resolution** (`kubernetes:///service.namespace:dispatch`)
- **RBAC permissions** included for endpoint watching (`get`, `list`, `watch`)
- **Consistent hash load balancing** for distributed request processing
- **Automatic enablement** with 2+ replicas (matches operator)
- **Comprehensive documentation** in production guide with verification commands
- **Operator parity** validated against [operator test configuration](https://github.com/authzed/spicedb-operator/blob/main/pkg/config/config_test.go)
- **Integration tests** validate 2-replica deployment with dispatch cluster enabled

### 🛡️ Security & High Availability
- **HA by default**: 2 replicas with dispatch enabled (matches operator)
- **Cloud workload identity support**: AWS EKS Pod Identity, GCP Workload Identity, Azure Workload Identity
- **Comprehensive TLS/mTLS**: gRPC, HTTP, dispatch cluster, and datastore connections
- **NetworkPolicy** templates for network isolation
- **PodDisruptionBudget** automatic enablement for HA deployments
- **Pod anti-affinity** and topology spread constraints for distributed deployments
- **Secure secret generation** with cryptographically secure `randBytes` (requires explicit opt-in)
- **Fail-fast validation** for missing secret configuration (prevents silent failures)

### 📦 Configuration Presets
Production-ready configurations matching operator deployment patterns:
- **`development.yaml`** - Single replica, memory datastore, minimal resources, auto-secret generation
- **`production-postgres.yaml`** - HPA (2-5 replicas), PostgreSQL, TLS, PDB, anti-affinity, topology spread
- **`production-cockroachdb.yaml`** - 2 replicas, CockroachDB, dispatch cluster with kubernetes:// discovery, mTLS

### 🔧 Advanced Features
- **Strategic merge patch system** for Deployment, Service, and Ingress customization
- **Operator-style configuration** with `operatorStyle.secretName` support
- **Automatic secret generation** with Helm lookup for seamless upgrades (opt-in)
- **Health monitoring script** (`scripts/status.sh`) for deployment validation
- **Configuration converter** scripts (Operator CR ↔ Helm values)

## 📊 Test Coverage

- **305 unit tests** across all components (helm-unittest)
- **113 migration hook tests** covering all edge cases
- **Integration tests** for HA deployments with dispatch cluster:
  - 2 replicas with dispatch enabled from start
  - kubernetes:// service discovery validation
  - RBAC permissions for endpoint watching
  - Rolling upgrades with dispatch cluster
  - Data persistence across upgrades
- **CI workflows**: Unit tests, preset validation, integration tests
- **Dry-run validation** in CI for all presets

### Reorganized Documentation Structure
Following best practices for clarity and maintainability:

```
docs/
├── quickstart.md                    # Get started in minutes
├── configuration/
│   └── presets.md                  # Preset documentation
├── migration/
│   ├── operator-comparison.md      # Feature comparison matrix
│   ├── helm-to-operator.md         # Migration guide
│   └── operator-to-helm.md         # Migration guide (35KB comprehensive)
├── operations/
│   └── status-script.md            # Monitoring and troubleshooting
├── guides/
│   ├── production.md               # Production deployment best practices
│   │                               # NEW: Dispatch Cluster section with kubernetes:// docs
│   ├── security.md                 # Security hardening guide
│   ├── troubleshooting.md          # Common issues and solutions
│   └── upgrading.md                # Version upgrade procedures
└── development/
    ├── testing.md                  # Testing guide (305+ tests)
    └── tech-debt.md                # Known issues and improvements
```

### New Documentation
- **Dispatch Cluster section** in production guide:
  - kubernetes:// service discovery explained
  - RBAC requirements documented
  - Port name vs port number resolution
  - How dispatch cluster works internally
  - Verification commands for troubleshooting
  - References to AuthZed blog and operator source
- **Comprehensive migration guides** (operator ↔ Helm)
- **Preset usage documentation** with examples
- **Security hardening guide** with TLS/mTLS setup
- **Production deployment guide** with HA best practices
- **Testing guide** documenting all 305+ tests
- **Status monitoring documentation** for operations

### Documentation Cleanup
- **YAML comment verbosity reduced by 59-69%** (~950 lines removed)
  - values.yaml: 965 → 294 lines (69% reduction)
  - Preset files: 465 → 189 lines (59% reduction)
  - Moved verbose examples to `examples/` directory
- **Enhanced dispatch documentation** in values.yaml:
  - kubernetes:// service discovery scheme documented
  - RBAC requirements noted
  - Automatic enablement conditions explained
- **Preset consolidation**: 4 presets → 3 presets
  - production-ha.yaml merged into production-postgres.yaml
  - Simplified deployment patterns
- **Cloud workload identity operator parity**
  - Changed from separate ServiceAccount manifests to serviceAccount.annotations
  - Matches operator pattern for AWS/GCP/Azure workload identity
- Moved all root documentation to `docs/` hierarchy (5 files)
- Removed 616 lines from supporting README files
- Focused tech-debt.md on active issues only (31% reduction)
- Updated all references to reflect 2-replica HA default with dispatch cluster

### New Design Principle: **Operator Parity**

Changed from "backward compatibility" to **"operator parity"** as primary design goal:

**Default Values (Operator Parity):**
- ✅ `replicaCount: 2` (operator default for HA)
- ✅ `dispatch.enabled: true` with kubernetes:// service discovery (operator default for persistent datastores)
- ✅ `migrations.logLevel: debug` (operator default for detailed logging)
- ✅ `config.autogenerateSecret: false` (security best practice, requires explicit opt-in)

**Dispatch Configuration:**
- ✅ kubernetes:// service discovery: `kubernetes:///service.namespace:dispatch`
- ✅ Port name resolution instead of port numbers
- ✅ RBAC with watch permission on endpoints
- ✅ Consistent hash load balancing

**Preserved for Development:**
- Development preset overrides to single replica, memory datastore, disabled dispatch
- Auto-secret generation enabled in development preset for convenience

### Critical Fixes
- **Migration hook secret references** - Fixed hooks failing when `autogenerateSecret=false` without `existingSecret`
- **Secret validation** - Added explicit fail-fast validation for missing secret configuration
- **Migration log level consistency** - Fixed default from "info" to "debug" for operator parity
- **CI preset validation** - Fixed preset tests failing due to secret auto-generation
- **Operator parity gaps** - Corrected 3 default value mismatches with operator
- **Dispatch service discovery** - Implemented kubernetes:// scheme matching operator
- **RBAC permissions** - Added watch permission for kubernetes:// endpoint discovery
- **Security improvements** - Changed to cryptographically secure `randBytes 32 | b64enc`

### Code Quality Fixes (PR Review)
- **Simplified template logic** - Reduced nested conditionals in secret.yaml
- **Added explanatory comments** - deployment.yaml now documents strategic merge patch system
- **Fixed test expectations** - Updated unit tests to expect kubernetes:// format instead of DNS
- **Consistent defaults** - Migration validation now uses "debug" matching migration job

### Integration Test Fixes
- Added secret availability checks to migration hooks
- Fixed CreateContainerConfigError in test scenarios
- Updated integration tests to validate 2-replica HA with dispatch
- All 113 migration hook tests now passing
- All 305 unit tests passing with new defaults

## 🚀 Release Automation

**Already Implemented** (corrected in tech-debt.md):
- ✅ Release Please workflow for automated releases
- ✅ Conventional Commits changelog generation
- ✅ Automatic Chart.yaml version bumping
- ✅ OCI registry publishing (ghcr.io/salekseev/helm-charts)
- ✅ GitHub release automation

### Before This PR
- Basic Helm chart with manual configuration
- Single replica default (not HA)
- Dispatch disabled by default
- DNS-based service discovery
- Scattered documentation
- Limited operator parity
- Silent failures on misconfiguration

### After This PR
- **True operator parity** (validated against source code)
- **HA by default** (2 replicas, dispatch enabled)
- **kubernetes:// service discovery** for dispatch cluster
- **3 production-ready presets** (simplified from 4)
- **Automated migration system** with tracking
- **Fail-fast validation** for configuration errors
- **Comprehensive documentation** (well-organized, 59-69% reduced verbosity, dispatch cluster section)
- **305 tests** covering all features including HA with dispatch
- **Production-grade security** (TLS/mTLS, NetworkPolicy, secure secrets)
- **Cloud provider integration** (workload identity via serviceAccount annotations)
- **Clean, maintainable codebase** with explicit validation

## 🔍 Tracked Improvements

Created Task Master tasks for remaining improvements:
- **Task #79**: Split `templates/_helpers.tpl` (638 lines → 7 focused files) - 2-3 hours
- **Task #80**: Split migration docs (1,400+ lines → subdirectories) - 1-2 hours
- **Task #78**: Operator-to-helm migration integration tests - 3-4 days

## ✅ Verification

- [x] All 305 unit tests passing
- [x] All 113 migration hook tests passing
- [x] Integration tests passing (HA with dispatch, kubernetes:// discovery, persistence, cleanup)
- [x] CI workflows passing (unit, integration, preset validation)
- [x] Template rendering successful for all presets
- [x] Helm lint passing (strict mode)
- [x] Documentation reviewed and organized
- [x] Dispatch cluster section added to production guide
- [x] kubernetes:// service discovery validated
- [x] RBAC permissions tested (endpoints watch)
- [x] Secret validation tested (fail-fast behavior confirmed)
- [x] Migration log level consistency verified
- [x] Code quality improvements applied (PR review comments)
- [x] YAML verbosity reduced (59-69% across files)
- [x] Preset consolidation complete (4 → 3 presets)
- [x] Cloud workload identity matches operator pattern
- [x] No functional regressions
- [x] Operator parity validated against source code
- [x] Development preset maintains dev-friendly defaults

## 🎓 Key Learnings

1. **Operator parity requires source code validation** - Documentation insufficient
2. **kubernetes:// service discovery** is operator standard for dispatch cluster
3. **Port name resolution** required for kubernetes:// scheme
4. **RBAC watch permission** essential for kuberesolver endpoint discovery
5. **Fail-fast validation** prevents confusing runtime errors
6. **Consistent defaults** across templates improve maintainability
7. **2 replicas provides basic HA** - Sufficient for most production deployments
8. **Configuration presets** provide better UX than complex default logic
9. **Comprehensive testing** (305 tests) catches edge cases early
10. **Terse documentation** improves readability (59-69% reduction achieved)
11. **Preset consolidation** reduces complexity (3 presets vs 4)
12. **serviceAccount annotations** match operator pattern for cloud workload identity
13. **Well-organized documentation** improves discoverability
14. **Explicit validation** is better than silent failures

### Option A: Accept New Defaults (Recommended for Production)
1. Review upgrade impact (scaling to 2 replicas with dispatch cluster)
2. Ensure cluster has resources for additional replica
3. Verify RBAC permissions are applied (automatic with chart)
4. Upgrade normally - dispatch with kubernetes:// discovery will auto-enable

**Benefits:**
- Basic HA out of the box
- Performance improvements with dispatch cluster
- Kubernetes-native service discovery
- Aligns with operator behavior
- Fail-fast validation prevents misconfigurations

### Option B: Maintain Current Behavior
Add to your `values.yaml`:
```yaml
replicaCount: 1
dispatch:
  enabled: false
migrations:
  logLevel: info
config:
  autogenerateSecret: true  # Not recommended for production
```

### Option C: Use Development Preset (Development Only)
```bash
helm upgrade spicedb . -f values-presets/development.yaml
```

## 🔗 References

- [SpiceDB Operator Source Code](https://github.com/authzed/spicedb-operator/blob/main/pkg/config/config.go)
- [Operator Dispatch Test Configuration](https://github.com/authzed/spicedb-operator/blob/main/pkg/config/config_test.go)
- [Consistent Hash Load Balancing for gRPC](https://authzed.com/blog/consistent-hash-load-balancing-grpc)
- [gRPC Kubernetes Resolver](https://github.com/sercand/kuberesolver)

## 🙏 Acknowledgments

- Operator source code analysis for true parity validation
- AuthZed team for kubernetes:// service discovery blog post
- PR review feedback on security best practices and code quality
- Copilot AI for identifying validation gaps and consistency issues
- spicedb-operator project for feature parity reference

---

**Ready for review!** All tests passing, documentation complete, **true operator parity achieved** with kubernetes:// service discovery and fail-fast validation. 🚀

**Note:** This PR introduces breaking changes and should be released as **v2.0.0** per semantic versioning.

### CODE REVIEW

File: charts/spicedb/MIGRATION_OPERATOR_TO_HELM.md [hunk 1]
```diff
@@ -0,0 +1,1395 @@
...
+# Save for conversion to Helm values
+cat spicedbcluster-backup.yaml
+
+# Document operator version
+kubectl get deployment -n spicedb-operator-system spicedb-operator-controller-manager -o jsonpath='{.spec.template.spec.containers[0].image}'
+```
+
+### 2. Backup Database
+
+Create a backup of your datastore **before** proceeding:
```
Comment (salekseev) [inline charts/spicedb/MIGRATION_OPERATOR_TO_HELM.md]: I believe SpiceDB has its own backup restore mechanism described at https://authzed.com/docs/spicedb/ops/data/migrations

**Review by salekseev — COMMENTED** [review]
I believe we should add a default service account to the chart to be able to use AWS EKS Pod Identity to connect to datastores as well as Cloud Workload Identity on other clouds.
